drv/lpc55-update-server/src/images.rs

```diff
@@ -26,79 +47,617 @@ extern "C" {
-                __IMAGE_B_BASE.as_ptr() as u32..__IMAGE_B_END.as_ptr() as u32,
-                __IMAGE_B_BASE.as_ptr() as u32..__IMAGE_B_END.as_ptr() as u32,
-            ),
-            (RotComponent::Stage0, SlotId::A) => (
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0_END.as_ptr() as u32,
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0_END.as_ptr() as u32,
-            ),
-            (RotComponent::Stage0, SlotId::B) => (
-                __IMAGE_STAGE0NEXT_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0NEXT_END.as_ptr() as u32,
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0_END.as_ptr() as u32,
-            ),
+/// Accessor keeps the implementation details of ImageAccess private
+enum Accessor<'a> {
+    // Flash driver, flash device range
+    Flash {
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        span: FlashRange,
+    },
+    Ram {
+        buffer: &'a [u8],
+        span: FlashRange,
+    },
+    // Hybrid is used for later implementation of rollback protection.
+    // The buffer is used in place of the beginning of the flash range.
+    Hybrid {
+        buffer: &'a [u8],
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        span: FlashRange,
+    },
+}
+
+impl Accessor<'_> {
+    fn at_runtime(&self) -> &Range<u32> {
+        match self {
+            Accessor::Flash { span, .. }
+            | Accessor::Ram { span, .. }
+            | Accessor::Hybrid { span, .. } => &span.at_runtime,
+        }
+    }
+}
+
+/// In addition to images that are located in their respective
+/// flash slots, the `update_server` needs to read data from
+/// complete and partial images in RAM or split between RAM
+/// and flash.
+/// The specific cases are when the
+///   - image is entirely in flash.
+///   - header block is in RAM with the remainder unavailable.
+///   - header block is in RAM with the remainder in flash.
+///   - entire image is in RAM (in the case of a cached Stage0 image).
+///
+/// Calls to methods use offsets into the image which is helpful
+/// when dealing with the offsets and sizes found in image headers
+/// and the caboose.
+pub struct ImageAccess<'a> {
+    accessor: Accessor<'a>,
+}
+
+impl ImageAccess<'_> {
+    pub fn new_flash<'a>(
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        component: RotComponent,
+        slot: SlotId,
+    ) -> ImageAccess<'a> {
+        let span = flash_range(component, slot);
+        ImageAccess {
+            accessor: Accessor::Flash { flash, span },
+        }
+    }
+
+    pub fn new_ram(
+        buffer: &[u8],
+        component: RotComponent,
+        slot: SlotId,
+    ) -> ImageAccess<'_> {
+        let span = flash_range(component, slot);
+        ImageAccess {
+            accessor: Accessor::Ram { buffer, span },
+        }
+    }
+
+    pub fn new_hybrid<'a>(
+        flash: &'a drv_lpc55_flash::Flash<'a>,
```

Comment:
Nit: if the order in memory is `[buffer, flash]`, then that should also be the argument order her (and constructor order below)